### PR TITLE
Add support for HA PROXY v1

### PR DESCRIPTION
**Pull request by cfcs**
Resolves hannesm#6

This commit adds a flag (`--haproxy1`) which causes tlstunnel to send
connection details to the destination upon connection.

This is useful if running `nginx`, `Varnish` or similar behind `tlstunnel` in order
to obtain the IP of the client.

Example of a TCP/IPv4 connection from 127.0.0.1:39837 to 127.0.0.1:4433:

```
PROXY TCP4 127.0.0.1 127.0.0.1 39837 4433
```

Relevant `nginx` documentation on how to use the PROXY protocol: https://www.nginx.com/resources/admin-guide/proxy-protocol/

### CODE REVIEW

**Comment by hannesm** [comment]
this looks great, thanks... I'll merge tomorrow after I tried it somehow

File: tlstunnel.ml [hunk 1]
```diff
@@ -77,6 +77,37 @@ module Fd_logger = struct
...
+   * This module only implements the protocol detailed in the
+   * "2.1. Human-readable header format (Version 1)" section.
+   *
+   * *)
+
+  let make_header socket =
+    (* basically it looks like:
+     * PROXY TCP4 SOURCEIP DESTIP SRCPORT DESTPORT\r\n *)
+    let own_sockaddr = getsockname socket in
+    let peer_sockaddr = getpeername socket in
```
Comment (hannesm) [inline tlstunnel.ml]: they should both be `Unix.getXname`...

**Comment by cfcs** [comment]
Example usage for testing:

Run `tlstunnel`:

```
user@localhost:~/tlsping (master)$ ../tlstunnel/tlstunnel.native --cert proxy.public.certificate --key proxy.secret.key --haproxy1
[2016-02-04T00:22:33Z] listener started on 0.0.0.0:4433, forwarding to 127.0.0.1:8080
[2016-02-04T00:22:52Z] 127.0.0.1:54508: connection established (TLS version 1.2, TLS_DHE_RSA_WITH_AES_256_CCM)
```

netcat listener:

```
user@localhost:~$ nc -v -l -p 8080
listening on [any] 8080 ...
connect to [127.0.0.1] from localhost [127.0.0.1] 50062
PROXY TCP4 127.0.0.1 127.0.0.1 54508 4433
```

connect using tlsclient:

```
user@localhost:~/tlsclient $ ./tlsclient.native localhost:4433
```

File: tlstunnel.ml [hunk 2]
```diff
@@ -77,6 +77,37 @@ module Fd_logger = struct
...
+    let get_addr_port = function
+      | Lwt_unix.ADDR_INET (inet_addr , port)
+        ->  (Unix.string_of_inet_addr inet_addr) , string_of_int port
+      | Lwt_unix.ADDR_UNIX _ -> failwith "TODO addr_unix" in
+    let peer_addr, peer_port = get_addr_port peer_sockaddr
+    and own_addr, own_port = get_addr_port own_sockaddr in
+    let header = String.concat " "
+    [ "PROXY" ; protocol_string ; peer_addr ; own_addr ; peer_port ; own_port ]
+    in
+    header ^ "\r\n"
```
Comment (avsm) [inline tlstunnel.ml:108]: this could be folded into the above `String.concat`

Comment (cfcs) [inline tlstunnel.ml:108]: Can you give an example?

Comment (hannesm) [inline tlstunnel.ml:108]: you could:

```
String.concat " "
   [ "PROXY" ; protocol_string ; peer_addr ; own_addr ; peer_port ; own_port ^ "\r\n" ]
```

I'm undecided which is more readable...

Comment (cfcs) [inline tlstunnel.ml:108]: Ah, right.
I'm fine with both, feel free to change it.

**Pull request merged**